packages/bff/src/services/purposeService.ts

```diff
@@ -4,16 +4,277 @@ import {
-import { BffAppContext } from "../utilities/context.js";
+import {
+  agreementApi,
+  bffApi,
+  catalogApi,
+  purposeApi,
+  tenantApi,
+} from "pagopa-interop-api-clients";
+import {
+  AgreementProcessClient,
+  AuthorizationProcessClient,
+  CatalogProcessClient,
+  PurposeProcessClient,
+  TenantProcessClient,
+} from "../providers/clientProvider.js";
+import {
+  agreementNotFound,
+  eServiceDescriptorNotFound,
+  eServiceNotFound,
+  purposeDraftVersionNotFound,
+  purposeNotFound,
+  tenantNotFound,
+} from "../model/domain/errors.js";
+import { BffAppContext, Headers } from "../utilities/context.js";
+import { getLatestAgreement } from "./agreementService.js";
+import { getAllClients } from "./authorizationService.js";
+
+export const getCurrentVersion = (
+  purposeVersions: purposeApi.PurposeVersion[]
+): purposeApi.PurposeVersion | undefined => {
+  const statesToExclude: purposeApi.PurposeVersionState[] = [
+    "WAITING_FOR_APPROVAL",
+    "REJECTED",
+  ];
+  // eslint-disable-next-line functional/immutable-data
+  return purposeVersions
+    .filter((v) => !statesToExclude.includes(v.state))
+    .sort(
+      (a, b) =>
+        new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
+    )
+    .pop();
+};
+
+// eslint-disable-next-line max-params
+async function getPurposes(
+  {
+    purposeClient,
+    eserviceClient,
+    tenantClient,
+    agreementClient,
+    authorizationClient,
+  }: {
+    purposeClient: PurposeProcessClient;
+    eserviceClient: CatalogProcessClient;
+    tenantClient: TenantProcessClient;
+    agreementClient: AgreementProcessClient;
+    authorizationClient: AuthorizationProcessClient;
+  },
+  requesterId: string,
+  filters: {
+    name?: string | undefined;
+    eservicesIds?: string[];
+    consumersIds?: string[];
+    producersIds?: string[];
+    states?: purposeApi.PurposeVersionState[];
+    excludeDraft?: boolean | undefined;
+  },
+  offset: number,
+  limit: number,
+  headers: Headers
+): Promise<bffApi.Purposes> {
+  const purposes = await purposeClient.getPurposes({
+    queries: {
+      ...filters,
+      limit,
+      offset,
+    },
+    withCredentials: true,
+    headers,
+  });
+
+  const eservices = await Promise.all(
+    [...new Set(purposes.results.map((p) => p.eserviceId))].map((id) =>
+      eserviceClient.getEServiceById({
+        params: {
+          eServiceId: id,
+        },
+        withCredentials: true,
+        headers,
+      })
+    )
+  );
+
+  const getTenant = async (id: string): Promise<tenantApi.Tenant> =>
+    tenantClient.tenant.getTenant({
+      params: {
+        id,
+      },
+      withCredentials: true,
+      headers,
+    });
+  const consumers = await Promise.all(
+    [...new Set(purposes.results.map((p) => p.consumerId))].map(getTenant)
+  );
+  const producers = await Promise.all(
+    [...new Set(eservices.map((e) => e.producerId))].map(getTenant)
+  );
+
+  const enhancePurpose = async (
```

Comment:
I would move it outside this `getPurposes` function since it's used also by other functions (you could also ignore and do it when actually needed)

Reply:
I prefer to move it when necessary